[Bugfix] Guard `_may_reorder_batch` for encoder-only models on CPU (#24319)

comsky opened the pull request:
Fixes AssertionError in `cpu_model_runner._may_reorder_batch` when running encoder-only models (e.g., `BAAI/bge-reranker-v2-m3`) on CPU backend. (refs [#24319](https://github.com/vllm-project/vllm/issues/24319))

Root cause:
- `kv_cache_groups == 1` but `attn_groups` is empty → direct access to `attn_groups[0][0].metadata_builder` raises `AssertionError`.

Fix:
- Add guard clauses to skip reorder when `attn_groups` is empty or not `TorchSDPAMetadataBuilderV1`.
- Safe for encoder-only/rerank models (reordering not beneficial).
- Decoder/attention-heavy models remain unaffected.

Tested with:
- `/v1/rerank` API on CPU backend using `BAAI/bge-reranker-v2-m3`.
- Verified no crash and valid rerank scores returned.

Happy to open further follow-up PRs if maintainers suggest adding unit tests for encoder-only models. 🙌

## Purpose

Fix a bug in the CPU backend where `_may_reorder_batch` incorrectly assumes the existence of `attn_groups`, leading to `AssertionError` when running encoder-only (reranker) models such as `BAAI/bge-reranker-v2-m3`.

## Test Plan

1. Build vLLM from this branch.
2. Launch the server with CPU backend and reranker model:
   ```bash
   python -m vllm.entrypoints.openai.api_server \
     --model BAAI/bge-reranker-v2-m3 \
     --task score \
     --runner pooling \
     --dtype float32 \
     --api-key token-abc123
   ```
3. Send a rerank request:
   ```bash
   curl -s http://127.0.0.1:8000/v1/rerank \
     -H "Authorization: Bearer token-abc123" \
     -H "Content-Type: application/json" \
     -d '{
       "model": "BAAI/bge-reranker-v2-m3",
       "query": "capital of France",
       "documents": [
         "The capital of Brazil is Brasilia.",
         "The capital of France is Paris."
       ]
     }'
   ```
4. Observe: No crash, valid scores returned.

## Test Result

* Before: Server crashes with `AssertionError` in `_may_reorder_batch`.
* After: No crash, rerank results returned successfully.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

comsky committed:
[Bugfix] Guard `_may_reorder_batch` for encoder-only models on CPU (#24319)

Fixes AssertionError in cpu_model_runner._may_reorder_batch when running
encoder-only models (e.g., BAAI/bge-reranker-v2-m3) on CPU backend.

Root cause:
- kv_cache_groups == 1 but attn_groups is empty → direct access to
  attn_groups[0][0].metadata_builder raises AssertionError.

Fix:
- Add guard clauses to skip reorder when attn_groups is empty or not
  TorchSDPAMetadataBuilderV1.
- Safe for encoder-only/rerank models (reordering not beneficial).
- Decoder/attention-heavy models remain unaffected.

Tested with:
- `/v1/rerank` API on CPU backend using BAAI/bge-reranker-v2-m3.
- Verified no crash and valid rerank scores returned.

Happy to open further follow-up PRs if maintainers suggest adding
unit tests for encoder-only models.

Signed-off-by: Remy <eunhwan.shin@dtonic.io>

bigPYJ1151 commented:
Checked with ```examples/offline_inference/basic/score.py```, I got some errors similar to #24407 and #24156

It should be a bug in ```unify_hybrid_kv_cache_specs``` code path. But looks like it didn't show in the online serving usage, weird.  

The root cause of entering ```unify_hybrid_kv_cache_specs``` is ```scheduler_config.disable_hybrid_kv_cache_manager``` is overwritten to ```True```. We should keep it as the default value ```False``` on CPU backend.
https://github.com/vllm-project/vllm/blob/60f0843ef8fb4b0c4e6788acc042873a0a2ea2a1/vllm/config/__init__.py#L3752-L3754

comsky committed:
[Bugfix] Disable hybrid KV cache manager overwrite and enable CPU tests for encoder-only models

- Set `disable_hybrid_kv_cache_manager = False` on CPU backend to avoid triggering the invalid path in `_may_reorder_batch`.
- Enable CPU test coverage for encoder-only (`BAAI/bge-base-en-v1.5`) and cross-encoder (`sentence-transformers/stsb-roberta-base-v2`) models.
- Together with the previous guard patch, this ensures rerank/score APIs run stably on CPU without AssertionError.

Signed-off-by: Remy <eunhwan.shin@dtonic.io>

bigPYJ1151 reviewed (commented):
Thanks for the fix, some further suggestions. 
BTW, please fix the failed DCO check https://github.com/vllm-project/vllm/pull/24348/checks?check_run_id=49841135329

comsky commented on vllm/config/__init__.py:
```diff
@@ -3743,7 +3743,7 @@ def __post_init__(self):
             if not (current_platform.is_cuda() or current_platform.is_rocm()):
                 # Hybrid KV cache manager is not supported on non-GPU platforms.
-                self.scheduler_config.disable_hybrid_kv_cache_manager = True
+                self.scheduler_config.disable_hybrid_kv_cache_manager = False
```
Thanks for catching that! 🙏 You’re right - modifying the global setting directly would have been risky. I’ll update the patch to add `current_platform.is_cpu()` to the condition instead.

comsky committed:
[Bugfix] Remove extra smoke test and refine CPU condition for hybrid KV cache manager

- Removed newly added test_cpu_score_rerank_smoke.py
- Updated config/__init__.py to refine the CPU handling logic by adding current_platform.is_cpu() to the condition

Signed-off-by: Remy <eunhwan.shin@dtonic.io>

bigPYJ1151 reviewed (approved):
LGTM, thanks for the fix! With this patch #24156 should also be resolved.

bigPYJ1151 committed:
Merge branch 'main' into bugfix/may-reorder-batch-cpu

comsky commented:
@bigPYJ1151 Thank you for your support! :)

comsky committed:
[Style] Apply pre-commit formatting fixes in config/__init__.py

Apply formatting changes suggested by pre-commit to fix CI check.

- Adjusted `if` condition layout in config/__init__.py
- No functional changes, formatting only

Signed-off-by: Remy <eunhwan.shin@dtonic.io>

comsky commented on vllm/v1/worker/cpu_model_runner.py:
```diff
@@ -55,11 +55,23 @@ def _may_reorder_batch(self, scheduler_output: "SchedulerOutput") -> None:
+        # may be empty or lack the expected metadata_builder.
+        # Without this check, accessing `attn_groups[0][0]` would trigger
+        # an AssertionError on CPU backend.
+        if not hasattr(self, "attn_groups") or not self.attn_groups:
```
Rather than special‑casing an “empty list,” this is a capability check: we reorder only when the first group’s `metadata_builder` is `TorchSDPAMetadataBuilderV1`; otherwise we return early. This prevents the CPU reranker crash we observed while keeping decoder/attention‑heavy paths unchanged, and the overhead is O(1).

bigPYJ1151 commented on vllm/config/__init__.py:
```diff
@@ -3749,7 +3749,8 @@ def __post_init__(self):
             # warning message here and will log it later.
-            if not (current_platform.is_cuda() or current_platform.is_rocm()):
+            if not (current_platform.is_cuda() or current_platform.is_rocm()
+                    or current_platform.is_cpu()):
```
Exactly. But for now only full attention is allowed on CPU backend and other attention specs will be aborted in earlier checks, it should be equivalent use or not use hybrid kv. At least existing cpu tests are passed. We can do more verification when enable more attention specs.

comsky commented:
@lianyiibo Adding an if kv_cache_spec_values at that location might only mask a call‑site issue rather than address it. It may be safer to gate the invocation upstream so unify_hybrid_kv_cache_specs runs only when the hybrid KV manager is enabled, the current runner/platform supports it, and specs are non‑empty. If a local guard is still desired, perhaps emit a warning instead of silently returning.
https://github.com/vllm-project/vllm/blob/b8a93076d36eff5cff8a89f99a7370d0cc6f0e98/vllm/v1/core/kv_cache_utils.py#L762

comsky committed:
Merge branch 'main' into bugfix/may-reorder-batch-cpu
